com.unity.ml-agents.extensions/Runtime/Sensors/GridSensor.cs

```diff
@@ -536,73 +553,112 @@ protected void ChannelsToTexture(int channelIndex, int numChannelsToAdd)
         /// <returns>A float[] containing all of the information collected from the gridsensor</returns>
         public float[] Perceive()
         {
+            if (m_CollidersPerCell == null)
+            {
+                return Array.Empty<float>();
+            }
             Reset();
             using (TimerStack.Instance.Scoped("GridSensor.Perceive"))
             {
-                // TODO: make these part of the class
-                Collider[] foundColliders = null;
-                Vector3 cellCenter = Vector3.zero;
+                var halfCellScale = new Vector3(CellScaleX / 2f, CellScaleY, CellScaleZ / 2f);
 
-                Vector3 halfCellScale = new Vector3(CellScaleX / 2f, CellScaleY, CellScaleZ / 2f);
-
-                for (int cellIndex = 0; cellIndex < NumCells; cellIndex++)
+                for (var cellIndex = 0; cellIndex < NumCells; cellIndex++)
                 {
+                    int numFound;
+                    Vector3 cellCenter;
                     if (RotateToAgent)
                     {
-                        cellCenter = transform.TransformPoint(CellPoints[cellIndex]);
-                        foundColliders = Physics.OverlapBox(cellCenter, halfCellScale, transform.rotation, ObserveMask);
+                        Transform transform1;
+                        cellCenter = (transform1 = transform).TransformPoint(CellPoints[cellIndex]);
+                        numFound = BufferResizingOverlapBoxNonAlloc(cellCenter, halfCellScale, transform1.rotation);
                     }
                     else
                     {
                         cellCenter = transform.position + CellPoints[cellIndex];
-                        foundColliders = Physics.OverlapBox(cellCenter, halfCellScale, Quaternion.identity, ObserveMask);
+                        numFound = BufferResizingOverlapBoxNonAlloc(cellCenter, halfCellScale, Quaternion.identity);
                     }
 
-                    if (foundColliders != null && foundColliders.Length > 0)
+                    if (numFound > 0)
                     {
-                        ParseColliders(foundColliders, cellIndex, cellCenter);
+                        ParseColliders(m_CollidersPerCell, numFound, cellIndex, cellCenter);
                     }
                 }
             }
-
             return m_PerceptionBuffer;
         }
 
+        /// <summary>
+        /// This method attempts to perform the Physics.OverlapBoxNonAlloc and will double the size of the Collider buffer
+        /// if the number of Colliders in the buffer after the call is equal to the length of the buffer.
+        /// </summary>
+        /// <param name="cellCenter"></param>
+        /// <param name="halfCellScale"></param>
+        /// <param name="rotation"></param>
+        /// <returns></returns>
+        int BufferResizingOverlapBoxNonAlloc(Vector3 cellCenter, Vector3 halfCellScale, Quaternion rotation)
+        {
+            int numFound;
+            while (true)
```

Comment:
```suggestion
            // Since we can only get a fixed number of results, requery
            // until we're sure we can hold them all (or until we hit the max size).
            while (true)
```